src/components/lib/kernel/cos_kernel_api.c

```diff
@@ -869,30 +985,62 @@ cos_comp_alloc_shared(struct cos_compinfo *ci, pgtblcap_t ptc, vaddr_t entry, st
 	return 0;
 }
 
+static pgtblcap_t
+__alloc_pgtbl_secondlvl(struct cos_compinfo *ci, vaddr_t heap_ptr)
+{
+	pgtblcap_t pte_cap;
+
+	pte_cap = __bump_mem_expand_intern(__compinfo_metacap(ci), ci->pgtbl_cap, heap_ptr, 0, 0);
+	assert(pte_cap);
+
+	/* 
+	 * this should only be called during component initialization
+	 * so I dont think this needs to happen atomically
+	 */
+	ci->vasrange_frontier[0] = cos_pgtbl_round_up_to_page(0, heap_ptr + PAGE_SIZE);
+
+	return pte_cap;
+}
 
 int
 cos_compinfo_alloc(struct cos_compinfo *ci, vaddr_t heap_ptr, capid_t cap_frontier, vaddr_t entry,
-                   struct cos_compinfo *ci_resources)
+                   struct cos_compinfo *ci_resources, prot_domain_t protdom)
 {
 	pgtblcap_t  ptc;
 	captblcap_t ctc;
 	compcap_t   compc;
-	int pgtbl_lvl;
+	vaddr_t     last_page;
+	int         pgtbl_lvl;
 
 	printd("cos_compinfo_alloc\n");
 
 	ptc = cos_pgtbl_alloc(ci_resources);
 	assert(ptc);
 	ctc = cos_captbl_alloc(ci_resources);
 	assert(ctc);
-	compc = cos_comp_alloc(ci_resources, ctc, ptc, entry);
+	compc = cos_comp_alloc(ci_resources, ctc, ptc, entry, protdom);
 	assert(compc);
 
 	cos_compinfo_init(ci, ptc, ctc, compc, heap_ptr, cap_frontier, ci_resources);
 
-	/* This is to make sure that "the address below vas_frontier has been allocated, follow the assumption we put in cos_vasfrontier_init()"*/
-	for (pgtbl_lvl = 0; pgtbl_lvl < COS_PGTBL_DEPTH - 1; pgtbl_lvl++) {
-		__bump_mem_expand_intern(ci, ptc, round_to_page(ci->vas_frontier - 1), 0, pgtbl_lvl);
+	ci->mi.second_lvl_pgtbl_cap  = __bump_mem_expand_intern(__compinfo_metacap(ci), ci->pgtbl_cap, heap_ptr, 0, 0);
```

Comment:
The change throughout this file to use the meta instead of the ci is HUGE. Perfectly fantastic if it is fixing a big bug, and is OK, but it is a big risk. Could break existing stuff. Just make sure to test it with enough setups that utilize the code.